File: Creational/Builder/TypeScript/Structure/Product.ts
/**
 * Product - Represents the complex object being built
 * Contains all the parts that can be assembled
 */
export class Product {
  private parts: string[] = [];

  /**
   * Adds a part to the product
   */
  addPart(part: string): void {
    this.parts.push(part);
  }

  /**
   * Lists all parts of the product
   */
  listParts(): string {
    return `Product parts: ${this.parts.join(", ")}`;
  }

  /**
   * Gets the number of parts
   */
  getPartCount(): number {
    return this.parts.length;
  }

  /**
   * Gets all parts as an array
   */
  getParts(): string[] {
    return [...this.parts];
  }

  /**
   * Clears all parts (for resetting)
   */
  clear(): void {
    this.parts = [];
  }

  /**
   * String representation of the product
   */
  toString(): string {
    return `Product(${this.parts.join(", ")})`;
  }
}

File: Creational/Builder/Python/Structure/product.py
"""
Product class that represents the complex object being built.
"""

class Product:
    """
    The Product class represents a complex object that is being constructed.
    It contains multiple parts that are added by different builders.
    """
    
    def __init__(self):
        self.parts = []
    
    def add_part(self, part: str):
        """Add a part to the product."""
        self.parts.append(part)
    
    def list_parts(self) -> str:
        """List all parts of the product."""
        return f"Product parts: {', '.join(self.parts)}"
    
    def __str__(self):
        return self.list_parts()
